### ISSUE TITLE
FastAPI / enable wiring for the DynamicContainer

### ISSUE THREAD

**Issue by nhdinh** [ets-labs/python-dependency-injector#427]
Hello, I am struggled to implements FastAPI endpoints to work with DynamicContainer, something like this:

```python
from dependency_injector import providers
from dependency_injector.containers import DynamicContainer
from dependency_injector.wiring import Provide
from fastapi import FastAPI, Depends, APIRouter


class ProductServices:
    def __init__(self):
        self.product_prefix = "P_"

    def get_product(self, product_id: int):
        return f"{self.product_prefix}{product_id}"


class RatingServices:
    def __init__(self, product_services=None):
        self.product_services = product_services

    def rate(self, product_id: int, point: int):
        product = self.product_services.get_product(product_id=product_id)
        return f"{product} has been rated {point} points"


router = APIRouter()


@router.get('/rate')
def rating(
        rating_services: RatingServices = Depends(Provide['rating_services'])
):
    if rating_services:
        return {
            'msg': rating_services.rate(10, 1)
        }


def populate_container(_container):
    _container.product_services = providers.Factory(ProductServices)
    _container.rating_services = providers.Factory(RatingServices,
                                                   product_services=_container.product_services)


def create_app():
    container = DynamicContainer()
    populate_container(container)

    app = FastAPI()
    app.container = container
    app.include_router(router=router)

    return app


app = create_app()
```

I can get an instance of `rating_services` after `populate_container` by calling `container.rating_services()` but cannot find a way to inject into FastAPI endpoint. Everytime making a request to `/rate` endpoint, FastAPI will response that: 

`
AttributeError: 'Provide' object has no attribute 'rate'
`

Please help or show me any workaround. Thanks alot

**Comment by rmk135** [comment]
Hi @nhdinh ,

There are two things missing in your example:

1) Add `@inject` decorator for the endpoint
2) Wire container with current module `container.wire(modules=[sys.modules[__name__]])`

Here is a working sample with comments:

```python
import sys  # <-- import sys

from dependency_injector import providers
from dependency_injector.containers import DynamicContainer
from dependency_injector.wiring import inject, Provide
from fastapi import FastAPI, Depends, APIRouter


class ProductServices:
    def __init__(self):
        self.product_prefix = "P_"

    def get_product(self, product_id: int):
        return f"{self.product_prefix}{product_id}"


class RatingServices:
    def __init__(self, product_services=None):
        self.product_services = product_services

    def rate(self, product_id: int, point: int):
        product = self.product_services.get_product(product_id=product_id)
        return f"{product} has been rated {point} points"


router = APIRouter()


@router.get('/rate')
@inject  # <-- add @inject decorator
def rating(
        rating_services: RatingServices = Depends(Provide['rating_services'])
):
    if rating_services:
        return {
            'msg': rating_services.rate(10, 1)
        }


def populate_container(_container):
    _container.product_services = providers.Factory(ProductServices)
    _container.rating_services = providers.Factory(RatingServices,
                                                   product_services=_container.product_services)


def create_app():
    container = DynamicContainer()
    populate_container(container)
    container.wire(modules=[sys.modules[__name__]])  # <-- wire container

    app = FastAPI()
    app.container = container
    app.include_router(router=router)

    return app


app = create_app()
```

**Comment by nhdinh** [comment]
> Hi @nhdinh ,
> 
> There are two things missing in your example:
> 
> 1. Add `@inject` decorator for the endpoint
> 2. Wire container with current module `container.wire(modules=[sys.modules[__name__]])`
> 
> Here is a working sample with comments:
> 
> ```python
> import sys  # <-- import sys
> 
> from dependency_injector import providers
> from dependency_injector.containers import DynamicContainer
> from dependency_injector.wiring import inject, Provide
> from fastapi import FastAPI, Depends, APIRouter
> 
> 
> class ProductServices:
>     def __init__(self):
>         self.product_prefix = "P_"
> 
>     def get_product(self, product_id: int):
>         return f"{self.product_prefix}{product_id}"
> 
> 
> class RatingServices:
>     def __init__(self, product_services=None):
>         self.product_services = product_services
> 
>     def rate(self, product_id: int, point: int):
>         product = self.product_services.get_product(product_id=product_id)
>         return f"{product} has been rated {point} points"
> 
> 
> router = APIRouter()
> 
> 
> @router.get('/rate')
> @inject  # <-- add @inject decorator
> def rating(
>         rating_services: RatingServices = Depends(Provide['rating_services'])
> ):
>     if rating_services:
>         return {
>             'msg': rating_services.rate(10, 1)
>         }
> 
> 
> def populate_container(_container):
>     _container.product_services = providers.Factory(ProductServices)
>     _container.rating_services = providers.Factory(RatingServices,
>                                                    product_services=_container.product_services)
> 
> 
> def create_app():
>     container = DynamicContainer()
>     populate_container(container)
>     container.wire(modules=[sys.modules[__name__]])  # <-- wire container
> 
>     app = FastAPI()
>     app.container = container
>     app.include_router(router=router)
> 
>     return app
> 
> 
> app = create_app()
> ```

Hi @rmk135 

Thanks for the reply. But just to mention here is that I am using `DynamicContainer` which don't allow me to `wire` the instance with the application module. Therefore, the `@inject` which I tried already also did not execute.

BTW, just to make clear why I want to use the `DynamicContainer` here. It's because the `RatingModule` and `ProductModule` which has the `RatingServices` and `ProductServices`, in respectively, is pluggable and need to lose coupling in this system. If the module manager, found any of these said module, it will attach the module container into the main container. That's why I cannot use `DeclarativeContainer` here.

But it's really hard to make the `DynamicContainer` working with FastAPI.

**Comment by rmk135** [comment]
> But just to mention here is that I am using DynamicContainer which don't allow me to wire the instance with the application module.

Why do you say that you can not wire dynamic container? That's how you do it:

```python
container = DynamicContainer()
populate_container(container)
container.wire(modules=[sys.modules[__name__]])  # <-- wire container
```

Am I missing anything?

**Comment by nhdinh** [comment]
> > But just to mention here is that I am using DynamicContainer which don't allow me to wire the instance with the application module.
> 
> Why do you say that you can not wire dynamic container? That's how you do it:
> 
> ```python
> container = DynamicContainer()
> populate_container(container)
> container.wire(modules=[sys.modules[__name__]])  # <-- wire container
> ```
> 
> Am I missing anything?

I've got an exception while make it wiring: "Can wire only an instance of the declarative container"

<img width="2672" alt="Screen Shot 2021-03-20 at 00 41 03" src="https://user-images.githubusercontent.com/5628052/111821567-64990300-8915-11eb-88db-191731f33de8.png">

**Comment by rmk135** [comment]
> BTW, just to make clear why I want to use the DynamicContainer here. It's because the RatingModule and ProductModule which has the RatingServices and ProductServices, in respectively, is pluggable and need to lose coupling in this system. If the module manager, found any of these said module, it will attach the module container into the main container. That's why I cannot use DeclarativeContainer here.

You could have a declarative container with a `Dependency` provider. This `Dependency` provider could be overridden in runtime.

```python
class Container(containers.DeclarativeContainer):

    database = providers.Dependency(instance_of=DbAdapter)

    user_service = providers.Factory(
        UserService,
        database=database,
    )


container = Container()
container.wire(...)
container.database.override(providers.Singleton(SqliteDbAdapter))  # database is permanently overridden, no need to re-wire.
```

**Comment by rmk135** [comment]
> I've got an exception while make it wiring: "Can wire only an instance of the declarative container"

Oh, gosh, I'm so sorry...

I tested the code from previous comment on a branch where this exception is already removed. I'll release that change and you would be able to wire dynamic container.

Apologies for a confusion.

**Comment by nhdinh** [comment]
> > I've got an exception while make it wiring: "Can wire only an instance of the declarative container"
> 
> Oh, gosh, I'm so sorry...
> 
> I tested the code from previous comment on a branch where this exception is already removed. I'll release that change and you would be able to wire dynamic container.
> 
> Apologies for a confusion.

That will be great. By the way, I just find a workaround for this by making a `DeclarativeContainer` with a factory method. The factory method will discover all the installed modules, collect their declared containers and mount into the ApplicationContainer which is created on-the-fly.

The ApplicationContainer can be wired easily.

**Comment by rmk135** [comment]
I've published version `4.30.0` that removes restriction on wiring a dynamic container. My apologies again for the confusion :) Closing the issue. Feel free to comment.
